src/Sarif.Multitool/Rules/SARIF1020.ReferToLatestSchema.cs

```
@@ -0,0 +1,54 @@
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Sarif.Multitool.Rules
+{
+    public class ReferToLatestSchema : SarifValidationSkimmerBase
+    {
+        private readonly MultiformatMessageString _fullDescription = new MultiformatMessageString
+        {
+            Text = RuleResources.SARIF1020_ReferToFinalSchema
+        };
+
+        public override MultiformatMessageString FullDescription => _fullDescription;
+
+        public override FailureLevel DefaultLevel => FailureLevel.Error;
+
+        /// <summary>
+        /// SARIF1020
+        /// </summary>
+        public override string Id => RuleId.ReferToLatestSchema;
+
+        protected override IEnumerable<string> MessageResourceNames => new string[]
+        {
+            nameof(RuleResources.SARIF1020_ReferenceToOldSchemaVersion),
+            nameof(RuleResources.SARIF1020_SchemaReferenceMissing)
+        };
+
+        protected override void Analyze(SarifLog log, string logPointer)
+        {
+            AnalyzeSchema(log.SchemaUri, logPointer);
+        }
+
+        private void AnalyzeSchema(Uri schemaUri, string pointer)
+        {
+            if (schemaUri == null)
+            {
+                LogResult(pointer, nameof(RuleResources.SARIF1020_SchemaReferenceMissing));
+                return;
+            }
+
+            if (!schemaUri.OriginalString.EndsWith(VersionConstants.StableSarifVersion)
+                && !schemaUri.OriginalString.EndsWith($"{VersionConstants.StableSarifVersion}.json")
+                && !schemaUri.OriginalString.EndsWith(VersionConstants.SchemaVersionAsPublishedToSchemaStoreOrg)
+                && !schemaUri.OriginalString.EndsWith($"{VersionConstants.SchemaVersionAsPublishedToSchemaStoreOrg}.json"))
+            {
+                LogResult(pointer, nameof(RuleResources.SARIF1020_ReferenceToOldSchemaVersion), schemaUri.OriginalString);
```

>pointer [](start = 26, length = 7)

Re-opening this comment so it's visible in CodeFlow on the latest iteration:

The `pointer` argument is a JSON Pointer string like `"/runs/0/results/0"` (which points to the first result in the first run in the log file). In this case, `pointer` will be `/` since we are analyzing the `SarifLog` object which represents the root of the log file.

When you are reporting that the `$schema` property is _missing_, then the location of the error really is `"/"`. But when `$schema` is present but invalid (because it is old), you actually want to report the error _at the location of the $schema property_.

The way you do that is to add a `using` directive for `Microsoft.Json.Pointer`. Then, the first argument to `LogResult` will be `pointer.AtProperty(SarifPropertyNames.Schema)`.